planner/optimize.go

```diff
@@ -118,109 +119,132 @@ func Optimize(ctx context.Context, sctx sessionctx.Context, node ast.Node, is in
 			return fp, fp.OutputNames(), nil
 		}
 	}
-
 	sctx.PrepareTSFuture(ctx)
 
-	bestPlan, names, _, err := optimize(ctx, sctx, node, is)
-	if err != nil {
-		return nil, nil, err
-	}
-	if !(sessVars.UsePlanBaselines || sessVars.EvolvePlanBaselines) {
-		return bestPlan, names, nil
-	}
+	useBinding := sessVars.UsePlanBaselines
 	stmtNode, ok := node.(ast.StmtNode)
 	if !ok {
-		return bestPlan, names, nil
-	}
-	bindRecord, scope, err := getBindRecord(sctx, stmtNode)
-	if err != nil {
-		return nil, nil, err
-	}
-	if bindRecord == nil {
-		return bestPlan, names, nil
-	}
-	if sctx.GetSessionVars().SelectLimit != math.MaxUint64 {
-		sctx.GetSessionVars().StmtCtx.AppendWarning(errors.New("sql_select_limit is set, so plan binding is not activated"))
-		return bestPlan, names, nil
-	}
-	err = setFoundInBinding(sctx, true)
-	if err != nil {
-		return nil, nil, err
-	}
-	bestPlanHint := plannercore.GenHintsFromPhysicalPlan(bestPlan)
-	if len(bindRecord.Bindings) > 0 {
-		orgBinding := bindRecord.Bindings[0] // the first is the original binding
-		for _, tbHint := range tableHints {  // consider table hints which contained by the original binding
-			if orgBinding.Hint.ContainTableHint(tbHint.HintName.String()) {
-				bestPlanHint = append(bestPlanHint, tbHint)
-			}
+		useBinding = false
+	}
+	if useBinding && sessVars.SelectLimit != math.MaxUint64 {
+		sessVars.StmtCtx.AppendWarning(errors.New("sql_select_limit is set, ignore SQL bindings"))
+		useBinding = false
+	}
+	var (
+		bindRecord *bindinfo.BindRecord
+		scope      string
+		err        error
+	)
+	if useBinding {
+		bindRecord, scope, err = getBindRecord(sctx, stmtNode)
+		if err != nil || bindRecord == nil || len(bindRecord.Bindings) == 0 {
```

Comment:
Should we handle the situation when `err  != nil`?

Reply:
I think we should not report error for user query because of bindings, maybe just a warning in the log is enough.